README.md
## Camera
<div align="center">
  <a href="http://quenty.github.io/NevermoreEngine/">
    <img src="https://github.com/Quenty/NevermoreEngine/actions/workflows/docs.yml/badge.svg" alt="Documentation status" />
  </a>
  <a href="https://discord.gg/mhtGUS8">
    <img src="https://img.shields.io/discord/385151591524597761?color=5865F2&label=discord&logo=discord&logoColor=white" alt="Discord" />
  </a>
  <a href="https://github.com/Quenty/NevermoreEngine/actions">
    <img src="https://github.com/Quenty/NevermoreEngine/actions/workflows/build.yml/badge.svg" alt="Build and release status" />
  </a>
</div>

CameraStackService provides a camera stack of objects that can report a camera state. This allows a composable camera system for a variety of situations that arise, and lets systems interop with each other

<div align="center"><a href="https://quenty.github.io/NevermoreEngine/api/CameraStackService">View docs →</a></div>

## Installation
```
npm install @bark_ing/camera --save
```

## Features

* Minor VR support
* Support for composable camera types
* Support for interpolation between camera nodes
* Support for touch and gamepad support and controls
* Support for camera shake

### Camera State
CameraState is an immutable camera system used for camera interpolation. Designed to make camera effects easy, it trades some efficiency and speed for making a not confusing camera system.

Camera state stores internal state of Coordinate frames as quaternions. This means that the classic lerp equation. Operations have been overridden to make this easier to work with.

Ok. It's not entirely immutable. They are kind of lazy, that is, reading data doesn't affect the state of them. There are no
required update loops or anything for the most part.

### Camera effect API
Current Camera Effects have the following API available. Adding new effects means that these two should follow the same API specifications

#### Add `+` operator
Returns a new Summed camera and adds the two effects together. This is used for combining effects

#### `.CameraState`
Indexing the CameraEffect should return the current camera state. This means that state is easy to index.

In previous versions, the following will also work. However, this is no longer true.

* `.State`
* `.CameraState`
* `.Camera`

This will return a `CameraState`

!!! warning
	Note that eventually we plan to deprecate everything but .CameraState

#### DefaultCamera
This class tracks the current camera ROBLOX uses and lets it maintain how it operates using a `BindToRenderStep` trick.

* Should call `BindToRenderStep` during construction
* Only one should exist at once per client (not enforced in code, however)

#### SummedCamera
This class takes two arguments and returns the summation of the two

* Arguments can be either `CameraState` or a CameraEffect, assuming the effect has a `CameraState` member

#### FadingCamera
This classes allows the effects of a camera to be faded / varied based upon a spring

* Starts at 0 percent effect

## Usage
Here is sample usage of using just a subcomponent. Recommendation is to use full camera stack service.

```lua
local RunService = game:GetService("RunService")
local Workspace = game:GetService("Workspace")

local FadeBetweenCamera3 = require(modules.FadeBetweenCamera3)
local CustomCameraEffect = require(modules.CustomCameraEffect)
local CameraState = require(modules.CameraState)

local defaultCamera = require(modules.DefaultCamera).new()
defaultCamera:BindToRenderStep() -- capture roblox camera automatically

local targetCamera = CustomCameraEffect.new(function()
  local target = CameraState.new()
  target.CFrame = CFrame.new(0, 100, 0)
  target.FieldOfView = 70

  return target
end)

local faded = FadeBetweenCamera3.new(defaultCamera, targetCamera)
faded.Speed = 5

RunService:BindToRenderStep("CameraStackUpdateInternal", Enum.RenderPriority.Camera.Value + 75, function()
  faded.CameraState:Set(Workspace.CurrentCamera)
end)

-- Input
local mouse = game.Players.LocalPlayer:GetMouse()
local visible = false
mouse.Button1Down:Connect(function()
  visible = not visible
  if visible then
    faded.Target = 1
  else
    faded.Target = 0
  end
end)

```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

### Features

* Add :SetAcceleration(acceleration) on GamepadRotateModel to allow for a more linear control of the camera ([3aec535](https://github.com/Quenty/NevermoreEngine/commit/3aec535dc88ed97ff4762e4b50d6e46d4c6c6ecb))


# [9.6.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.5.0...@quenty/camera@9.6.0) (2022-12-06)

**Note:** Version bump only for package @quenty/camera





# [9.5.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.4.0...@quenty/camera@9.5.0) (2022-11-19)

**Note:** Version bump only for package @quenty/camera





# [9.4.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.3.1...@quenty/camera@9.4.0) (2022-11-12)

**Note:** Version bump only for package @quenty/camera





## [9.3.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.3.0...@quenty/camera@9.3.1) (2022-11-04)

**Note:** Version bump only for package @quenty/camera





# [9.3.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.2.0...@quenty/camera@9.3.0) (2022-10-30)

**Note:** Version bump only for package @quenty/camera





# [9.2.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.1.1...@quenty/camera@9.2.0) (2022-10-23)


### Features

* Move CameraStack to its own reusable class ([7020502](https://github.com/Quenty/NevermoreEngine/commit/70205024acb508552c0c922a0a8119a27b7f38cd))





## [9.1.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.1.0...@quenty/camera@9.1.1) (2022-10-16)


### Bug Fixes

* CameraStack ends profile correctly if disabled ([cdca1bd](https://github.com/Quenty/NevermoreEngine/commit/cdca1bdb4dfcc7da576687114d2343568db131ab))





# [9.1.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@9.0.0...@quenty/camera@9.1.0) (2022-10-11)

**Note:** Version bump only for package @quenty/camera





# [9.0.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@8.1.0...@quenty/camera@9.0.0) (2022-09-27)

**Note:** Version bump only for package @quenty/camera





# [8.1.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@8.0.1...@quenty/camera@8.1.0) (2022-08-22)

**Note:** Version bump only for package @quenty/camera





## [8.0.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@8.0.0...@quenty/camera@8.0.1) (2022-08-16)

**Note:** Version bump only for package @quenty/camera





# [8.0.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@7.3.0...@quenty/camera@8.0.0) (2022-08-14)


### Features

* Add ServiceName to most services for faster debugging ([39fc3f4](https://github.com/Quenty/NevermoreEngine/commit/39fc3f4f2beb92fff49b2264424e07af7907324e))





# [7.3.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@7.2.0...@quenty/camera@7.3.0) (2022-07-31)

**Note:** Version bump only for package @quenty/camera





# [7.2.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@7.1.0...@quenty/camera@7.2.0) (2022-07-02)


### Bug Fixes

* Can clean up services properly ([eb45e03](https://github.com/Quenty/NevermoreEngine/commit/eb45e03ce2897b18f1ae460974bf2bbb9e27cb97))





# [7.1.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@7.0.0...@quenty/camera@7.1.0) (2022-06-21)


### Bug Fixes

* Delay CameraStackService starting until start method so that configuration has time to be set ([02d68d1](https://github.com/Quenty/NevermoreEngine/commit/02d68d1d8fbba5b9685c29c503520796c55fa3fa))





# [7.0.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@6.3.0...@quenty/camera@7.0.0) (2022-05-21)

**Note:** Version bump only for package @quenty/camera





# [6.3.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@6.2.0...@quenty/camera@6.3.0) (2022-03-27)

**Note:** Version bump only for package @quenty/camera





# [6.2.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@6.1.0...@quenty/camera@6.2.0) (2022-03-20)

**Note:** Version bump only for package @quenty/camera





# [6.1.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@6.0.0...@quenty/camera@6.1.0) (2022-03-10)

**Note:** Version bump only for package @quenty/camera





# [6.0.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.3.0...@quenty/camera@6.0.0) (2022-03-06)

**Note:** Version bump only for package @quenty/camera





# [5.3.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.2.1...@quenty/camera@5.3.0) (2022-01-17)

**Note:** Version bump only for package @quenty/camera





## [5.2.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.2.0...@quenty/camera@5.2.1) (2022-01-16)

**Note:** Version bump only for package @quenty/camera





# [5.2.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.1.1...@quenty/camera@5.2.0) (2022-01-07)

**Note:** Version bump only for package @quenty/camera





## [5.1.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.1.0...@quenty/camera@5.1.1) (2022-01-06)

**Note:** Version bump only for package @quenty/camera





# [5.1.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.0.2...@quenty/camera@5.1.0) (2022-01-03)

**Note:** Version bump only for package @quenty/camera





## [5.0.2](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.0.1...@quenty/camera@5.0.2) (2021-12-30)

**Note:** Version bump only for package @quenty/camera





## [5.0.1](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@5.0.0...@quenty/camera@5.0.1) (2021-12-30)

**Note:** Version bump only for package @quenty/camera





# [5.0.0](https://github.com/Quenty/NevermoreEngine/compare/@quenty/camera@4.4.0...@quenty/camera@5.0.0) (2021-12-22)


src/Client/Effects/ImpulseCamera.lua
--[=[
	Add another layer of effects over any other camera by allowing an "impulse"
	to be applied. Good for shockwaves, camera shake, and recoil.

	@class ImpulseCamera
]=]

local require = require(script.Parent.loader).load(script)

local CameraState = require("CameraState")
local Spring = require("Spring")
local SummedCamera = require("SummedCamera")

local ImpulseCamera = {}
ImpulseCamera.ClassName = "ImpulseCamera"

function ImpulseCamera.new()
	local self = setmetatable({
		_spring = Spring.new(Vector3.new(0, 0, 0))
	}, ImpulseCamera)

	self._spring.Damper = 0.5
	self._spring.Speed = 20

	return self
end

--[=[
	Applies an impulse to the camera, shaking it!
	@param velocity Vector3
]=]
function ImpulseCamera:Impulse(velocity)
	assert(typeof(velocity) == "Vector3", "Bad velocity")

	if velocity.Magnitude > 10 then
		velocity = velocity.Unit * 10
	end
	
	self._spring:Impulse(velocity)
end

function ImpulseCamera:__add(other)
	return SummedCamera.new(self, other)
end

function ImpulseCamera:__newindex(index, value)
	if index == "Damper" then
		assert(type(value) == "number", "Bad value")
		self._spring.Damper = value
	elseif index == "Speed" then
		assert(type(value) == "number", "Bad value")
		self._spring.Speed = value
	else
		error(("%q is not a valid member of impulse camera"):format(tostring(index)))
	end
end

--[=[
	The current state.
	@readonly
	@prop CameraState CameraState
	@within DefaultCamera
]=]
function ImpulseCamera:__index(index)
	if index == "CameraState" then
		local newState = CameraState.new()

		local position = self._spring.Value
		newState.CFrame = CFrame.Angles(0, position.y, 0)
			* CFrame.Angles(position.x, 0, 0)
			* CFrame.Angles(0, 0, position.z)

		return newState
	elseif index == "Damper" then
		return self._spring.Damper
	elseif index == "Speed" then
		return self._spring.Speed
	elseif index == "Spring" then
		return self._spring
	else
		return ImpulseCamera[index]
	end
end

return ImpulseCamera

src/Client/CameraStackService.lua
--[=[
	Holds camera states and allows for the last camera state to be retrieved. Also
	initializes an impulse and default camera as the bottom of the stack. Is a singleton.

	@class CameraStackService
]=]

local require = require(script.Parent.loader).load(script)

local RunService = game:GetService("RunService")
local Workspace = game:GetService("Workspace")
local HttpService = game:GetService("HttpService")

local DefaultCamera = require("DefaultCamera")
local ImpulseCamera = require("ImpulseCamera")
local ServiceBag = require("ServiceBag")
local Maid = require("Maid")
local CameraStack = require("CameraStack")

assert(RunService:IsClient(), "[CameraStackService] - Only require CameraStackService on client")

local CameraStackService = {}
CameraStackService.ServiceName = "CameraStackService"
CameraStackService._cameraStack = nil
CameraStackService._defaultCamera = nil
CameraStackService._rawDefaultCamera = nil
CameraStackService._impulseCamera = nil
CameraStackService._maid = nil
CameraStackService._doNotUseDefaultCamera = nil

--[=[
	Initializes a new camera stack. Should be done via the ServiceBag.
	@param serviceBag ServiceBag
]=]
function CameraStackService:Init(serviceBag)
	assert(ServiceBag.isServiceBag(serviceBag), "Not a valid service bag")

	CameraStackService._maid = Maid.new()
	CameraStackService._key = HttpService:GenerateGUID(false)

	CameraStackService._cameraStack = CameraStack.new()

	-- Initialize default cameras
	CameraStackService._rawDefaultCamera = DefaultCamera.new()
	CameraStackService._maid:GiveTask(CameraStackService._rawDefaultCamera)

	CameraStackService._impulseCamera = ImpulseCamera.new()
	CameraStackService._defaultCamera = (CameraStackService._rawDefaultCamera + CameraStackService._impulseCamera):SetMode("Relative")

	-- Add camera to stack
	CameraStackService:Add(CameraStackService._defaultCamera)

	RunService:BindToRenderStep("CameraStackUpdateInternal" .. CameraStackService._key, Enum.RenderPriority.Camera.Value + 75, function()
		debug.profilebegin("camerastackservice")

		local state = CameraStackService:GetTopState()
		if state then
			state:Set(Workspace.CurrentCamera)
		end

		debug.profileend()
	end)

	CameraStackService._maid:GiveTask(function()
		RunService:UnbindFromRenderStep("CameraStackUpdateInternal" .. CameraStackService._key)
	end)
end

function CameraStackService:Start()
	CameraStackService._started = true

	-- TODO: Allow rebinding
	if CameraStackService._doNotUseDefaultCamera then
		Workspace.CurrentCamera.CameraType = Enum.CameraType.Scriptable

		-- TODO: Handle camera deleted too!
		Workspace.CurrentCamera:GetPropertyChangedSignal("CameraType"):Connect(function()
			Workspace.CurrentCamera.CameraType = Enum.CameraType.Scriptable
		end)
	else
		CameraStackService._rawDefaultCamera:BindToRenderStep()
	end
end

--[=[
	Prevents the default camera from being used
	@param doNotUseDefaultCamera boolean
]=]
function CameraStackService:SetDoNotUseDefaultCamera(doNotUseDefaultCamera)
	assert(not CameraStackService._started, "Already started")

	CameraStackService._doNotUseDefaultCamera = doNotUseDefaultCamera
end

--[=[
	Pushes a disable state onto the camera stack
	@return function -- Function to cancel disable
]=]
function CameraStackService:PushDisable()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:PushDisable()
end

--[=[
	Outputs the camera stack. Intended for diagnostics.
]=]
function CameraStackService:PrintCameraStack()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:PrintCameraStack()
end

--[=[
	Returns the default camera
	@return SummedCamera -- DefaultCamera + ImpulseCamera
]=]
function CameraStackService:GetDefaultCamera()
	assert(CameraStackService._defaultCamera, "Not initialized")

	return CameraStackService._defaultCamera
end

--[=[
	Returns the impulse camera. Useful for adding camera shake.

	Shaking the camera:
	```lua
	CameraStackService._cameraStackService:GetImpulseCamera():Impulse(Vector3.new(0.25, 0, 0.25*(math.random()-0.5)))
	```

	You can also sum the impulse camera into another effect to layer the shake on top of the effect
	as desired.

	```lua
	-- Adding global custom camera shake to a custom camera effect
	local customCameraEffect = ...
	return (customCameraEffect + CameraStackService._cameraStackService:GetImpulseCamera()):SetMode("Relative")
	```

	@return ImpulseCamera
]=]
function CameraStackService:GetImpulseCamera()
	assert(CameraStackService._impulseCamera, "Not initialized")

	return CameraStackService._impulseCamera
end

--[=[
	Returns the default camera without any impulse cameras
	@return DefaultCamera
]=]
function CameraStackService:GetRawDefaultCamera()
	assert(CameraStackService._rawDefaultCamera, "Not initialized")

	return CameraStackService._rawDefaultCamera
end

--[=[
	Gets the camera current on the top of the stack
	@return CameraEffect
]=]
function CameraStackService:GetTopCamera()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetTopCamera()
end

--[=[
	Retrieves the top state off the stack at this time
	@return CameraState?
]=]
function CameraStackService:GetTopState()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetTopState()
end

--[=[
	Returns a new camera state that retrieves the state below its set state.

	@return CustomCameraEffect -- Effect below
	@return (CameraState) -> () -- Function to set the state
]=]
function CameraStackService:GetNewStateBelow()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetNewStateBelow()
end

--[=[
	Retrieves the index of a state
	@param state CameraEffect
	@return number? -- index

]=]
function CameraStackService:GetIndex(state)
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetIndex(state)
end

--[=[
	Returns the current stack.

	:::warning
	Do not modify this stack, this is the raw memory of the stack
	:::

	@return { CameraState<T> }
]=]
function CameraStackService:GetRawStack()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetRawStack()
end

--[=[
	Gets the current camera stack

	@return CameraStack
]=]
function CameraStackService:GetCameraStack()
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:GetStack()
end

--[=[
	Removes the state from the stack
	@param state CameraState
]=]
function CameraStackService:Remove(state)
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:Remove(state)
end

--[=[
	Adds the state from the stack
	@param state CameraState
]=]
function CameraStackService:Add(state)
	assert(CameraStackService._cameraStack, "Not initialized")

	return CameraStackService._cameraStack:Add(state)
end

function CameraStackService:Destroy()
	CameraStackService._maid:DoCleaning()
end

return CameraStackService